x/ecocredit/migrations/v3/core.go

```diff
@@ -0,0 +1,404 @@
+	for {
+		var classInfo ClassInfo
+		if _, err := classItr.LoadNext(&classInfo); err != nil {
+			if orm.ErrIteratorDone.Is(err) {
+				break
+			}
+			return err
+		}
+
+		admin, err := sdk.AccAddressFromBech32(classInfo.Admin)
+		if err != nil {
+			return err
+		}
+		dest := api.ClassInfo{
+			Name:       classInfo.ClassId,
+			Admin:      admin,
+			Metadata:   string(classInfo.Metadata),
+			CreditType: classInfo.CreditType.Abbreviation,
+		}
+		classID, err := ss.ClassInfoTable().InsertReturningID(ctx, &dest)
+		if err != nil {
+			return err
+		}
+		classIDNameMap[classID] = classInfo.ClassId
+		classNameIDMap[classInfo.ClassId] = classID
+
+		// migrate class issuers to ORM v1
+		for _, issuer := range classInfo.Issuers {
+			addr, err := sdk.AccAddressFromBech32(issuer)
+			if err != nil {
+				return err
+			}
+
+			if err := ss.ClassIssuerTable().Insert(ctx, &api.ClassIssuer{
+				ClassId: classID,
+				Issuer:  addr,
+			}); err != nil {
+				return err
+			}
+		}
+
+		// delete class info from old store
+		if err := classInfoTable.Delete(sdkCtx, &ClassInfo{ClassId: classInfo.ClassId}); err != nil {
+			return err
+		}
+	}
+
+	// migrate credit type sequence to ORM v1
+	cItr, err := creditTypeSeqTable.PrefixScan(sdkCtx, nil, nil)
+	if err != nil {
+		return err
+	}
+	defer cItr.Close()
+
+	for {
+		var ctype CreditTypeSeq
+		if _, err := cItr.LoadNext(&ctype); err != nil {
+			if orm.ErrIteratorDone.Is(err) {
+				break
+			}
+
+			return err
+		}
+		if err := ss.ClassSequenceTable().Save(ctx, &api.ClassSequence{
+			CreditType:  ctype.Abbreviation,
+			NextClassId: ctype.SeqNumber,
+		}); err != nil {
+			return err
+		}
+
+		// delete credit type sequence from old store
+		if err := creditTypeSeqTable.Delete(sdkCtx, &ctype); err != nil {
+			return err
+		}
+	}
+
+	// migrate credit batches to ORM v1 and create projects for existing credit classes
+	batchIDsMap := make(map[string]batchMapT) // map of a batch denom to batch-id and amount cancelled
+	batchSeqMap := make(map[uint64]uint64)    // map of a project-id to batch sequence
+	projectSeqMap := make(map[uint64]uint64)  // map of a credit classID to project sequence
+	batchItr, err := batchInfoTable.PrefixScan(sdkCtx, nil, nil)
+	if err != nil {
+		return err
+	}
+	defer batchItr.Close()
+
+	for {
+		var batchInfo BatchInfo
+		if _, err := batchItr.LoadNext(&batchInfo); err != nil {
+			if orm.ErrIteratorDone.Is(err) {
+				break
+			}
+			return err
+		}
+
+		admin, err := sdk.AccAddressFromBech32(batchInfo.Issuer)
+		if err != nil {
+			return err
+		}
+		pItr, err := ss.ProjectInfoTable().List(ctx, api.ProjectInfoAdminIndexKey{}.WithAdmin(admin.Bytes()))
+		if err != nil {
+			return err
+		}
+
+		projectExists := false
+		var projectID uint64
+		for pItr.Next() {
+			pInfo, err := pItr.Value()
+			if err != nil {
+				return err
+			}
+
+			if pInfo.ClassId == classNameIDMap[batchInfo.ClassId] {
+				projectExists = true
+				projectID = pInfo.Id
+				break
+			}
```

Comment:
This comment was from before the product call. Pending a response from @clevinson and @aaronc.

Reply:
After further discussing the @clevinson on the product call, we decided this would be the best approach to support both mainnet and redwood and then if we needed to perform any additional operations for mainnet (#980), we would handle those updates in the upgrade handler. Once this change is made and the related test is updated, this pull request has my approval.

Reply:
updated [3cd4bb1](https://github.com/regen-network/regen-ledger/pull/957/commits/3cd4bb11930724de96e955890ddf9eef0d5261d1#diff-f31c5489648094d493af1a8cfe6393dd8292a82f4e0902c5b981d3126bf36aeaR165)